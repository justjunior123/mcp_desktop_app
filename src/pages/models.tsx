import React from 'react';
import { ModelListClient } from '../components/models/ModelListClient.tsx';
import { getApiUrl } from '../config/api.ts';
import type { OllamaModelDetails } from '../services/ollama/ModelManager.ts';

interface ModelsPageProps {
  initialModels: OllamaModelDetails[];
}

export async function getServerSideProps() {
  const res = await fetch(getApiUrl('models'));
  if (!res.ok) return { notFound: true };
  const data: { models: OllamaModelDetails[] } = await res.json();
  return { props: { initialModels: data.models } };
}

export default function ModelsPage({ initialModels }: ModelsPageProps) {
  return (
    <div className="container mx-auto px-4 py-8">
      <h1 className="text-3xl font-bold mb-8">Model Management</h1>
      <ModelListClient initialModels={initialModels} />
    </div>
  );
} 