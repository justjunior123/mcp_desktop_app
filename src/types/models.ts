export interface Model {
  name: string;
  modelId: string;
  model?: string;
  modified_at?: string;
  size?: number;
  digest?: string;
  details?: {
    parent_model?: string;
    format?: string;
    family?: string;
    families?: string[];
    parameter_size?: string;
    quantization_level?: string;
  };
  status?: string;
  progress?: number;
} 