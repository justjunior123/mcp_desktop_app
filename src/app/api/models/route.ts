import { NextRequest, NextResponse } from 'next/server';
import { ModelManager } from '@/services/ollama/ModelManager';
import { DatabaseService } from '@/services/database/DatabaseService';
import { OllamaService } from '@/services/ollama/OllamaService';
import { logger } from '@/services/logging';
import { z } from 'zod';

const modelManager = new ModelManager(
  new DatabaseService(),
  new OllamaService()
);

// Input validation schemas
const postModelSchema = z.object({
  modelName: z.string().min(1, 'Model name is required')
});

// Error response helper
function handleError(error: unknown, context: string) {
  const errorObj = {
    message: error instanceof Error ? error.message : String(error),
    stack: error instanceof Error ? error.stack : undefined,
    timestamp: new Date().toISOString()
  };

  logger.error(`Error in ${context}:`, errorObj);

  if (error instanceof z.ZodError) {
    return NextResponse.json({
      error: 'Validation error',
      details: error.errors
    }, { status: 400 });
  }

  return NextResponse.json({
    error: errorObj.message,
    context
  }, { status: 500 });
}

export async function GET() {
  try {
    logger.info('Fetching models list');
    const models = await modelManager.listModelsWithDetails();
    logger.info('Successfully fetched models list', { count: models.length });
    return NextResponse.json({ models });
  } catch (error) {
    return handleError(error, 'GET /api/models');
  }
}

export async function POST(request: NextRequest) {
  try {
    const body = await request.json();
    logger.info('Received model pull request', { body });

    const { modelName } = postModelSchema.parse(body);
    
    logger.info('Starting model pull', { modelName });
    const model = await modelManager.pullModel(modelName);
    logger.info('Successfully pulled model', { modelName, model });
    
    return NextResponse.json({ model });
  } catch (error) {
    return handleError(error, 'POST /api/models');
  }
}

export async function DELETE(request: NextRequest) {
  try {
    const { searchParams } = new URL(request.url);
    const modelId = searchParams.get('modelId');
    logger.info('Received model delete request', { modelId });

    if (!modelId) {
      logger.warn('Model delete request missing modelId');
      return NextResponse.json({ error: 'Model ID is required' }, { status: 400 });
    }

    logger.info('Starting model deletion', { modelId });
    await modelManager.deleteModel(modelId);
    logger.info('Successfully deleted model', { modelId });
    
    return NextResponse.json({ success: true });
  } catch (error) {
    return handleError(error, 'DELETE /api/models');
  }
} 