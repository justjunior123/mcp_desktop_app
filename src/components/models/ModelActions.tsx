import React, { useState } from 'react';
import type { OllamaModelDetails } from '@services/ollama/types.ts';

interface ModelActionsProps {
  model: OllamaModelDetails;
  onPull: (modelName: string) => void;
  onDelete: (modelId: string) => void;
  onConfigure: (modelId: string) => void;
}

export const ModelActions: React.FC<ModelActionsProps> = ({
  model,
  onPull,
  onDelete,
  onConfigure,
}) => {
  const [progress, setProgress] = useState<number>(0);
  const [status, setStatus] = useState<string>('');

  const handlePull = async (modelName: string) => {
    setStatus('DOWNLOADING');
    setProgress(0);

    try {
      const response = await fetch(`/api/models/${modelName}/pull`);
      const reader = response.body?.getReader();
      const decoder = new TextDecoder();

      if (!reader) {
        throw new Error('Failed to start model pull');
      }

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        const chunk = decoder.decode(value);
        const lines = chunk.split('\n').filter(Boolean);

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            const data = JSON.parse(line.slice(6));
            setStatus(data.status);
            if (data.progress !== undefined) {
              setProgress(data.progress);
            }
          }
        }
      }

      onPull(modelName);
    } catch (error) {
      setStatus('ERROR');
      console.error('Error pulling model:', error);
    }
  };

  return (
    <div className="flex gap-2">
      {model.status === 'NOT_DOWNLOADED' && (
        <button
          onClick={() => handlePull(model.name)}
          className="px-4 py-2 bg-blue-500 text-white rounded hover:bg-blue-600"
          disabled={status === 'DOWNLOADING'}
        >
          Pull Model
        </button>
      )}
      {model.status === 'AVAILABLE' && (
        <>
          <button
            onClick={() => onConfigure(model.id)}
            className="px-4 py-2 bg-green-500 text-white rounded hover:bg-green-600"
          >
            Configure
          </button>
          <button
            onClick={() => onDelete(model.id)}
            className="px-4 py-2 bg-red-500 text-white rounded hover:bg-red-600"
          >
            Delete
          </button>
        </>
      )}
      {status === 'DOWNLOADING' && (
        <div className="flex items-center gap-2">
          <div className="w-32 bg-gray-200 rounded-full h-2.5">
            <div
              className="bg-blue-500 h-2.5 rounded-full transition-all duration-300"
              style={{ width: `${progress}%` }}
            ></div>
          </div>
          <span>{Math.round(progress)}%</span>
        </div>
      )}
      {status === 'ERROR' && (
        <div className="text-red-500">
          Error: {model.error || 'Unknown error'}
        </div>
      )}
    </div>
  );
}; 