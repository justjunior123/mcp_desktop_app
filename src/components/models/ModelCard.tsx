import React from 'react';
import { Model } from '@prisma/client';
import { OllamaModelDetails } from '../../services/ollama/ModelManager';
import { formatBytes } from '../../lib/utils';

type ModelStatus = 'installed' | 'not_installed' | 'downloading' | 'error' | 'deleting';

interface ModelCardProps {
  model: Model & { ollamaDetails?: OllamaModelDetails | null };
  onPull: (modelName: string) => void;
  onDelete: (modelId: string) => void;
  onConfigure: (modelId: string) => void;
  onViewDetails: (modelId: string) => void;
}

export const ModelCard: React.FC<ModelCardProps> = ({
  model,
  onPull,
  onDelete,
  onConfigure,
  onViewDetails,
}) => {
  const { id, name, status, ollamaDetails } = model;
  
  // Format model size if available
  const formattedSize = ollamaDetails?.size 
    ? formatBytes(Number(ollamaDetails.size)) 
    : 'Unknown size';
  
  // Determine the status color and text
  const getStatusInfo = (status: string) => {
    switch (status) {
      case 'installed':
        return { color: 'bg-green-500', text: 'Installed' };
      case 'not_installed':
        return { color: 'bg-gray-500', text: 'Not Installed' };
      case 'downloading':
        return { color: 'bg-blue-500', text: 'Downloading' };
      case 'error':
        return { color: 'bg-red-500', text: 'Error' };
      case 'deleting':
        return { color: 'bg-yellow-500', text: 'Deleting' };
      default:
        return { color: 'bg-gray-500', text: status };
    }
  };
  
  const { color, text } = getStatusInfo(status);
  
  // Get parameter size and quantization info
  const modelInfo = ollamaDetails ? (
    <div className="text-sm text-gray-500">
      {ollamaDetails.parameterSize && <span>{ollamaDetails.parameterSize} • </span>}
      {ollamaDetails.quantizationLevel && <span>{ollamaDetails.quantizationLevel} • </span>}
      <span>{formattedSize}</span>
    </div>
  ) : (
    <div className="text-sm text-gray-500">No details available</div>
  );
  
  // Calculate download progress if applicable
  const downloadProgress = ollamaDetails?.downloadProgress || 0;
  const showProgress = status === 'downloading' && downloadProgress < 100;

  return (
    <div className="bg-white dark:bg-gray-800 shadow rounded-lg p-4 hover:shadow-md transition-shadow">
      <div className="flex justify-between items-start">
        <div>
          <h3 className="text-lg font-medium text-gray-900 dark:text-white">{name}</h3>
          {modelInfo}
        </div>
        <div className={`${color} text-white text-xs px-2 py-1 rounded-full`}>
          {text}
        </div>
      </div>
      
      {/* Download progress */}
      {showProgress && (
        <div className="mt-3">
          <div className="w-full bg-gray-200 rounded-full h-2.5 dark:bg-gray-700">
            <div 
              className="bg-blue-600 h-2.5 rounded-full" 
              style={{ width: `${downloadProgress}%` }}
            ></div>
          </div>
          <div className="text-xs text-gray-500 mt-1 text-right">
            {downloadProgress.toFixed(0)}%
          </div>
        </div>
      )}
      
      {/* Error message */}
      {status === 'error' && ollamaDetails?.errorMessage && (
        <div className="mt-2 text-sm text-red-500">
          {ollamaDetails.errorMessage}
        </div>
      )}
      
      {/* Action buttons */}
      <div className="mt-4 flex flex-wrap gap-2">
        {status === 'not_installed' ? (
          <button
            onClick={() => onPull(name)}
            className="bg-blue-500 hover:bg-blue-600 text-white px-3 py-1 text-sm rounded"
          >
            Pull Model
          </button>
        ) : status === 'installed' ? (
          <>
            <button
              onClick={() => onConfigure(id)}
              className="bg-purple-500 hover:bg-purple-600 text-white px-3 py-1 text-sm rounded"
            >
              Configure
            </button>
            <button
              onClick={() => onDelete(id)}
              className="bg-red-500 hover:bg-red-600 text-white px-3 py-1 text-sm rounded"
            >
              Delete
            </button>
          </>
        ) : null}
        
        <button
          onClick={() => onViewDetails(id)}
          className="bg-gray-200 hover:bg-gray-300 dark:bg-gray-700 dark:hover:bg-gray-600 text-gray-800 dark:text-white px-3 py-1 text-sm rounded"
        >
          Details
        </button>
      </div>
    </div>
  );
}; 