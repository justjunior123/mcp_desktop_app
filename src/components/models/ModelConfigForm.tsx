import React, { useState } from 'react';
import { OllamaModelDetails } from '../../services/ollama/ModelManager';

interface ModelConfigFormProps {
  model: OllamaModelDetails;
  onSave: (modelId: string, parameters: string) => void;
  onBack: () => void;
}

export const ModelConfigForm: React.FC<ModelConfigFormProps> = ({
  model,
  onSave,
  onBack,
}) => {
  const [config, setConfig] = useState({
    temperature: model.configuration?.temperature ?? 0.7,
    topP: model.configuration?.topP ?? 0.9,
    topK: model.configuration?.topK ?? 40,
    repeatPenalty: model.configuration?.repeatPenalty ?? 1.1,
    presencePenalty: model.configuration?.presencePenalty ?? 0,
    frequencyPenalty: model.configuration?.frequencyPenalty ?? 0,
    contextWindow: model.configuration?.contextWindow ?? 4096,
    systemPrompt: model.configuration?.systemPrompt ?? '',
  });

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    onSave(model.id, JSON.stringify(config));
  };

  return (
    <div className="bg-white dark:bg-gray-800 rounded-lg shadow-md p-6">
      <div className="flex justify-between items-center mb-6">
        <h2 className="text-2xl font-bold">Configure {model.name}</h2>
        <button
          onClick={onBack}
          className="px-4 py-2 bg-gray-500 text-white rounded hover:bg-gray-600"
        >
          Back
        </button>
      </div>

      <form onSubmit={handleSubmit} className="space-y-6">
        <div className="grid grid-cols-1 md:grid-cols-2 gap-6">
          <div className="space-y-4">
            <div>
              <label className="block text-sm font-medium mb-1">Temperature</label>
              <input
                type="number"
                step="0.1"
                min="0"
                max="2"
                value={config.temperature}
                onChange={(e) => setConfig({ ...config, temperature: parseFloat(e.target.value) })}
                className="w-full p-2 border rounded focus:ring-2 focus:ring-blue-500"
              />
            </div>

            <div>
              <label className="block text-sm font-medium mb-1">Top P</label>
              <input
                type="number"
                step="0.1"
                min="0"
                max="1"
                value={config.topP}
                onChange={(e) => setConfig({ ...config, topP: parseFloat(e.target.value) })}
                className="w-full p-2 border rounded focus:ring-2 focus:ring-blue-500"
              />
            </div>

            <div>
              <label className="block text-sm font-medium mb-1">Top K</label>
              <input
                type="number"
                step="1"
                min="1"
                value={config.topK}
                onChange={(e) => setConfig({ ...config, topK: parseInt(e.target.value) })}
                className="w-full p-2 border rounded focus:ring-2 focus:ring-blue-500"
              />
            </div>

            <div>
              <label className="block text-sm font-medium mb-1">Repeat Penalty</label>
              <input
                type="number"
                step="0.1"
                min="1"
                value={config.repeatPenalty}
                onChange={(e) => setConfig({ ...config, repeatPenalty: parseFloat(e.target.value) })}
                className="w-full p-2 border rounded focus:ring-2 focus:ring-blue-500"
              />
            </div>
          </div>

          <div className="space-y-4">
            <div>
              <label className="block text-sm font-medium mb-1">Presence Penalty</label>
              <input
                type="number"
                step="0.1"
                min="-2"
                max="2"
                value={config.presencePenalty}
                onChange={(e) => setConfig({ ...config, presencePenalty: parseFloat(e.target.value) })}
                className="w-full p-2 border rounded focus:ring-2 focus:ring-blue-500"
              />
            </div>

            <div>
              <label className="block text-sm font-medium mb-1">Frequency Penalty</label>
              <input
                type="number"
                step="0.1"
                min="-2"
                max="2"
                value={config.frequencyPenalty}
                onChange={(e) => setConfig({ ...config, frequencyPenalty: parseFloat(e.target.value) })}
                className="w-full p-2 border rounded focus:ring-2 focus:ring-blue-500"
              />
            </div>

            <div>
              <label className="block text-sm font-medium mb-1">Context Window</label>
              <input
                type="number"
                step="512"
                min="512"
                value={config.contextWindow}
                onChange={(e) => setConfig({ ...config, contextWindow: parseInt(e.target.value) })}
                className="w-full p-2 border rounded focus:ring-2 focus:ring-blue-500"
              />
            </div>
          </div>
        </div>

        <div>
          <label className="block text-sm font-medium mb-1">System Prompt</label>
          <textarea
            value={config.systemPrompt}
            onChange={(e) => setConfig({ ...config, systemPrompt: e.target.value })}
            rows={4}
            className="w-full p-2 border rounded focus:ring-2 focus:ring-blue-500"
          />
        </div>

        <div className="flex justify-end">
          <button
            type="submit"
            className="px-6 py-2 bg-blue-500 text-white rounded hover:bg-blue-600"
          >
            Save Configuration
          </button>
        </div>
      </form>
    </div>
  );
}; 