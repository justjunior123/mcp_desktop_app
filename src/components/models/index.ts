export { ModelCard } from './ModelCard';
export { ModelList } from './ModelList';
export { ModelDetails } from './ModelDetails';
export { ModelConfigForm } from './ModelConfigForm';