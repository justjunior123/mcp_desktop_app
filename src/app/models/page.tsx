'use client';

import React, { useState, useEffect, Suspense } from 'react';
import { ModelList } from '@/components/models/ModelList';
import { ModelDetails } from '../../components/models/ModelDetails';
import { ModelConfigForm } from '../../components/models/ModelConfigForm';
import { useWebSocket } from '../../lib/hooks/useWebSocket';
import { Model } from '@prisma/client';
import { OllamaModelDetails } from '../../services/ollama/ModelManager';
import { ModelActions } from '@/components/models/ModelActions';
import { ServerStatusCard } from '@/components/ServerStatusCard';
import { getApiUrl, getWsUrl } from '@/config/api';

type ModelWithDetails = Model & { ollamaDetails?: OllamaModelDetails | null };
type ViewMode = 'list' | 'details' | 'config';

const ModelsPage: React.FC = () => {
  const [models, setModels] = useState<ModelWithDetails[]>([]);
  const [loading, setLoading] = useState(true);
  const [selectedModel, setSelectedModel] = useState<ModelWithDetails | null>(null);
  const [viewMode, setViewMode] = useState<ViewMode>('list');

  // WebSocket connection for real-time updates
  const { sendMessage } = useWebSocket(getWsUrl('ws'), {
    onMessage: (message) => {
      if (message.type === 'initialStatus') {
        // Initial status with model list
        setModels(message.payload.models);
        setLoading(false);
      } else if (message.type === 'modelStatusUpdate') {
        // Update model status in the list
        updateModelStatus(message.payload);
      } else if (message.type === 'modelDetails') {
        // Detailed model information
        setSelectedModel(message.payload);
        setViewMode('details');
      } else if (message.type === 'parametersSaved') {
        // Confirmation of saved parameters
        fetchModels();
        setViewMode('details');
      } else if (message.type === 'error') {
        // Error message
        console.error(message.payload.message);
      }
    },
    onConnect: () => {
      console.log('Connected to WebSocket server');
      // Request initial model list
      sendMessage({ type: 'refreshModels' });
    },
    onDisconnect: () => {
      console.log('Disconnected from WebSocket server');
    },
  });

  // Update model status in the list
  const updateModelStatus = (update: { modelId: string; status: string; downloadProgress?: number; error?: string }) => {
    setModels((prevModels) => {
      return prevModels.map((model) => {
        if (model.id === update.modelId) {
          return {
            ...model,
            status: update.status,
            ollamaDetails: model.ollamaDetails
              ? {
                  ...model.ollamaDetails,
                  downloadProgress: update.downloadProgress || model.ollamaDetails.downloadProgress,
                  downloadStatus: update.status,
                  errorMessage: update.error || model.ollamaDetails.errorMessage,
                }
              : null,
          };
        }
        return model;
      });
    });

    // Also update the selected model if it's affected
    if (selectedModel && selectedModel.id === update.modelId) {
      setSelectedModel((prev) => {
        if (!prev) return null;
        return {
          ...prev,
          status: update.status,
          ollamaDetails: prev.ollamaDetails
            ? {
                ...prev.ollamaDetails,
                downloadProgress: update.downloadProgress || prev.ollamaDetails.downloadProgress,
                downloadStatus: update.status,
                errorMessage: update.error || prev.ollamaDetails.errorMessage,
              }
            : null,
        };
      });
    }
  };

  // Fetch models from the API
  const fetchModels = async () => {
    try {
      const response = await fetch(getApiUrl('models'));
      if (!response.ok) {
        throw new Error(`Error fetching models: ${response.statusText}`);
      }
      const data = await response.json();
      setModels(data.models || []);
      setLoading(false);
    } catch (error) {
      console.error(error instanceof Error ? error.message : 'Unknown error fetching models');
      setLoading(false);
    }
  };

  // Pull a model from Ollama
  const handlePullModel = async (modelName: string) => {
    try {
      const response = await fetch(`${getApiUrl('models')}/pull`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ modelName }),
      });

      if (!response.ok) {
        throw new Error(`Error pulling model: ${response.statusText}`);
      }

      // The WebSocket will update the model status
    } catch (error) {
      console.error(error instanceof Error ? error.message : 'Unknown error pulling model');
    }
  };

  // Delete a model from Ollama
  const handleDeleteModel = async (modelId: string) => {
    if (!confirm('Are you sure you want to delete this model?')) {
      return;
    }

    try {
      const response = await fetch(`${getApiUrl('models')}/${modelId}`, {
        method: 'DELETE',
      });

      if (!response.ok) {
        throw new Error(`Error deleting model: ${response.statusText}`);
      }

      // If the deleted model is the selected one, go back to list
      if (selectedModel && selectedModel.id === modelId) {
        setSelectedModel(null);
        setViewMode('list');
      }

      // The WebSocket will update the model list
    } catch (error) {
      console.error(error instanceof Error ? error.message : 'Unknown error deleting model');
    }
  };

  // View model details
  const handleViewDetails = async (modelId: string) => {
    try {
      const response = await fetch(`${getApiUrl('models')}/${modelId}`);
      if (!response.ok) {
        throw new Error(`Error fetching model details: ${response.statusText}`);
      }
      const data = await response.json();
      setSelectedModel(data.model || null);
      setViewMode('details');
    } catch (error) {
      console.error(error instanceof Error ? error.message : 'Unknown error fetching model details');
    }
  };

  // Configure model parameters
  const handleConfigureModel = (modelId: string) => {
    // Find the model in our current list
    const model = models.find((m) => m.id === modelId) || selectedModel;
    if (model) {
      setSelectedModel(model);
      setViewMode('config');
    }
  };

  // Save model parameters
  const handleSaveParameters = async (modelId: string, parameters: string) => {
    try {
      const response = await fetch(`${getApiUrl('models')}/${modelId}/parameters`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ parameters }),
      });

      if (!response.ok) {
        throw new Error(`Error saving parameters: ${response.statusText}`);
      }

      // The WebSocket will send a parametersSaved message
    } catch (error) {
      console.error(error instanceof Error ? error.message : 'Unknown error saving parameters');
    }
  };

  // Render the appropriate content based on view mode
  const renderContent = () => {
    if (loading) {
      return <div>Loading...</div>;
    }

    switch (viewMode) {
      case 'details':
        return selectedModel ? (
          <ModelDetails
            model={selectedModel}
            onBack={() => {
              setSelectedModel(null);
              setViewMode('list');
            }}
            onConfigure={handleConfigureModel}
            onDelete={handleDeleteModel}
          />
        ) : null;

      case 'config':
        return selectedModel ? (
          <ModelConfigForm
            model={selectedModel}
            onSave={handleSaveParameters}
            onCancel={() => setViewMode('details')}
          />
        ) : null;

      case 'list':
      default:
        return (
          <>
            <ModelList
              models={models}
              onViewDetails={handleViewDetails}
              onConfigure={handleConfigureModel}
              onDelete={handleDeleteModel}
            />
            <ModelActions onPullModel={handlePullModel} />
            <ServerStatusCard />
          </>
        );
    }
  };

  return <div className="container mx-auto px-4 py-8">{renderContent()}</div>;
};

export default ModelsPage; 