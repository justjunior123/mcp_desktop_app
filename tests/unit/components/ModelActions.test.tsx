import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { ModelActions } from '@/components/models/ModelActions';
import type { OllamaModelDetails } from '@/services/ollama/types';

describe('ModelActions Component', () => {
  const mockModel: OllamaModelDetails = {
    id: '1',
    name: 'test-model',
    status: 'AVAILABLE' as const,
    error: null,
    size: BigInt(1000000),
    digest: 'sha256:test-digest',
    modified_at: new Date().toISOString(),
    parameters: {
      temperature: 0.7,
      top_p: 0.9,
      top_k: 40,
      repeat_penalty: 1.1
    }
  };

  const mockOnPull = jest.fn();
  const mockOnDelete = jest.fn();
  const mockOnConfigure = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders pull button when model is not downloaded', () => {
    const notDownloadedModel = { ...mockModel, status: 'NOT_DOWNLOADED' as const };
    render(
      <ModelActions
        model={notDownloadedModel}
        onPull={mockOnPull}
        onDelete={mockOnDelete}
        onConfigure={mockOnConfigure}
      />
    );

    const pullButton = screen.getByText('Pull Model');
    expect(pullButton).toBeInTheDocument();
    fireEvent.click(pullButton);
    expect(mockOnPull).toHaveBeenCalledWith('test-model');
  });

  it('renders configure and delete buttons when model is available', () => {
    render(
      <ModelActions
        model={mockModel}
        onPull={mockOnPull}
        onDelete={mockOnDelete}
        onConfigure={mockOnConfigure}
      />
    );

    const configureButton = screen.getByText('Configure');
    const deleteButton = screen.getByText('Delete');

    expect(configureButton).toBeInTheDocument();
    expect(deleteButton).toBeInTheDocument();

    fireEvent.click(configureButton);
    expect(mockOnConfigure).toHaveBeenCalledWith('1');

    fireEvent.click(deleteButton);
    expect(mockOnDelete).toHaveBeenCalledWith('1');
  });

  it('shows loading state when model is downloading', () => {
    const downloadingModel = { ...mockModel, status: 'DOWNLOADING' as const };
    render(
      <ModelActions
        model={downloadingModel}
        onPull={mockOnPull}
        onDelete={mockOnDelete}
        onConfigure={mockOnConfigure}
      />
    );

    expect(screen.getByText('Downloading...')).toBeInTheDocument();
    expect(screen.getByRole('status')).toHaveClass('animate-spin');
  });

  it('displays error message when model has error', () => {
    const errorModel = { ...mockModel, status: 'ERROR' as const, error: 'Test error' };
    render(
      <ModelActions
        model={errorModel}
        onPull={mockOnPull}
        onDelete={mockOnDelete}
        onConfigure={mockOnConfigure}
      />
    );

    expect(screen.getByText('Error: Test error')).toBeInTheDocument();
  });
}); 