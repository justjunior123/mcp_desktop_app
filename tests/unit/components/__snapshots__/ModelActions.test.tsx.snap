// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelActions Component renders correctly when model has error 1`] = `
<div>
  <div
    class="flex gap-2"
  >
    <div
      class="text-red-500"
    >
      Error: Test error
    </div>
  </div>
</div>
`;

exports[`ModelActions Component renders correctly when model is available 1`] = `
<div>
  <div
    class="flex gap-2"
  >
    <button
      class="px-4 py-2 bg-green-500 text-white rounded hover:bg-green-600"
      onClick={[Function]}
    >
      Configure
    </button>
    <button
      class="px-4 py-2 bg-red-500 text-white rounded hover:bg-red-600"
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;

exports[`ModelActions Component renders correctly when model is downloading 1`] = `
<div>
  <div
    class="flex gap-2"
  >
    <div
      class="flex items-center gap-2"
    >
      <div
        class="animate-spin rounded-full h-5 w-5 border-b-2 border-blue-500"
        role="status"
      />
      <span>
        Downloading...
      </span>
    </div>
  </div>
</div>
`;

exports[`ModelActions Component renders correctly when model is not downloaded 1`] = `
<div>
  <div
    class="flex gap-2"
  >
    <button
      class="px-4 py-2 bg-blue-500 text-white rounded hover:bg-blue-600"
      onClick={[Function]}
    >
      Pull Model
    </button>
  </div>
</div>
`;
